### PR TITLE
[ADP-3305] Add send faucet end-point to local cluster

### PR DESCRIPTION
**Pull request by paolino**
- [x] Add a send faucet assets end-point to the local cluster http service

ADP-3305

### CODE REVIEW

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/ApiT.hs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    )
+where
+
+import Prelude
+
+import GHC.Generics
+    ( Generic
+    )
+
+-- | A newtype wrapper to avoid orphan instances
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/ApiT.hs]:
```suggestion
-- | A value of type 'ApiT'@ a@ is equivalent to a value of type @a@.
-- However, the type 'ApiT'@ a@ has a custom JSON encoding that
-- is independent of the original definition of @a@.
--
-- In other words, the 'ApiT' type constructor allows us to separate the
-- concern of presenting @a@ in a HTTP/JSON interface from
-- the definition of @a@ as a data type.
```

If 'ApiT' is defined in a separate module, then the instances will be orphan instances, so the comment is no longer true. 😅 This comment should be a better description of the purpose.

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Application/Server.hs [hunk 2]
```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Cardano.Wallet.Launch.Cluster.Monitoring.Http.Application.Server
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Application/Server.hs]:
```suggestion
module Cardano.Wallet.Launch.Cluster.Http.Faucet.Server
```

This place in the module hierarchy would make more sense to me? 🤔 

I would definitely remove the `Monitoring` prefix.

I don't mind `Application`, though the term is a bit generic. `Faucet` would be more specific.

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Client.hs [hunk 3]
```diff
@@ -1,56 +1,45 @@
...
-{-# LANGUAGE GADTs #-}
-{-# LANGUAGE NamedFieldPuns #-}
-{-# LANGUAGE NumericUnderscores #-}
-{-# LANGUAGE RankNTypes #-}
-{-# LANGUAGE RecordWildCards #-}
-{-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE LambdaCase #-}
 
 module Cardano.Wallet.Launch.Cluster.Monitoring.Http.Client
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Client.hs]:
```suggestion
module Cardano.Wallet.Launch.Cluster.Http.Client
```

This module combines both the `Application`/`Faucet` and the `Monitoring`, doesn't it? 🤔

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Control/API.hs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Wallet.Launch.Cluster.Monitoring.Http.Control.API
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Control/API.hs]:
```suggestion
module Cardano.Wallet.Launch.Cluster.Http.Monitoring.API
```

? 🤔

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Control/Client.hs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Wallet.Launch.Cluster.Monitoring.Http.Control.Client
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/Control/Client.hs]:
```suggestion
module Cardano.Wallet.Launch.Cluster.Http.Monitoring.Client
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/SendFaucetAssets.hs [hunk 6]
```diff
@@ -0,0 +1,256 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Cardano.Wallet.Launch.Cluster.Monitoring.Http.SendFaucetAssets
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/SendFaucetAssets.hs]:
```suggestion
module Cardano.Wallet.Launch.Cluster.Http.SendFaucetAssets
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/SendFaucetAssets.hs [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+import qualified Cardano.Address as Addr
+
+-- | Payload to send assets to a list of addresses
+data SendFaucetAssets = SendFaucetAssets
+    { batchSize :: Int
+    -- ^ batch size
+    , assets :: [(Address, (TokenBundle, [(String, String)]))]
+    -- ^ List of addresses and the assets to send to each address
+    }
+    deriving stock (Eq, Show)
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Http/SendFaucetAssets.hs]:
```suggestion
type AssetMetadata = [(String, String)]

-- | Payload to send assets to a list of addresses
data SendFaucetAssets = SendFaucetAssets
    { batchSize :: Int
    -- ^ batch size
    , assets :: [(Address, (TokenBundle, AssetMetadata))]
    -- ^ List of addresses and the assets to send to each address
    }
    deriving stock (Eq, Show)
```

Type synonym for documentation purposes, makes it easier to understand what this list is supposed to be.

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Monitor.hs [hunk 8]
```diff
@@ -98,18 +120,51 @@ timedMonitor initialState = do
...
+withMonitoringServer network conn clusterConfig tr MonitorConfiguration{..} = do
     monitor <- liftIO $ withTracingState timedMonitor monitorInitialState
     port <- liftIO $ maybe getRandomPort pure monitorPort
     liftIO $ traceWith tr $ MsgHttpMonitoringPort port
-    withHttpServer port $ mkHandlers $ rmap History monitor
+    withHttpServer
+        network
+        port
+        (mkControlHandlers $ rmap History monitor)
+        (mkApplicationHandlers conn clusterConfig)
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Monitor.hs]: But the `mkApplicationHandlers` are no longer part of the `Monitoring` concern proper, are they? 🤔 Maybe move `withMonitoringServer` to the testing code?

**Review by HeinrichApfelmus — COMMENTED** [review]
Thank you! 😊

The code itself looks good to me, but I feel that the separation of concerns is confused. 🤔 I would have expected that the HTTP API for the local cluster sits under

`Cardano.Wallet.Launch.Cluster.Http.*`

and that the monitoring part is grouped under

`Cardano.Wallet.Launch.Cluster.Http.Monitoring.*`

while the faucet part is grouped under

`Cardano.Wallet.Launch.Cluster.Http.Faucet.*`
or ``Cardano.Wallet.Launch.Cluster.Http.Application.*`

It's true that the local cluster is a group of nodes that are running, but at the same time, I think it's also fair to say that "this group has an HTTP interface" (this interface happens to be served by the `local-cluster` executable) as opposed to "this group can be monitored by an additional process, and the monitor to has an HTTP interface". I think it's conceptually simpler to attribute the HTTP interface to "the cluster".

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Monitor.hs (hunk 8 above)
Comment (paolino) [reply lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Monitoring/Monitor.hs]: The HTTP server is the same, so we have to compose them. The api backed by `ApplicationHandlers` are intended to be offered within the local-cluster features.

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        batchSize <- o .: "batch-size"
+        assets <- o .: "assets" >>= parseAssets (sNetworkId @n)
+        pure $ WithNetwork $ SendFaucetAssets{batchSize, assets}
+
+--- assets parsing/rendering ---------------------------------------------------
+
+parseAssets
+    :: SNetworkId n
+    -> Value
+    -> Parser [(Address, (TokenBundle, [(String, String)]))]
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
    -> Parser [(Address, (TokenBundle, AssetMetadata))]
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+parseAssets n = parseJSON >=> mapM (parseAsset n)
+
+parseAsset
+    :: SNetworkId n
+    -> Value
+    -> Parser (Address, (TokenBundle, [(String, String)]))
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
    -> Parser (Address, (TokenBundle, AssetMetadata))
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+parseAsset n = withObject "Asset" $ \o -> do
+    addr <- o .: "address" >>= parseAddress n
+    bundle <- o .: "bundle" >>= parseBundle
+    metadata <- o .: "metadata" >>= parseMetadata
+    pure (addr, (bundle, metadata))
+
+renderAssets
+    :: SNetworkId n
+    -> [(Address, (TokenBundle, [(String, String)]))]
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
    -> [(Address, (TokenBundle, AssetMetadata))]
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+    -> Value
+renderAssets n = toJSON . map (renderAsset n)
+
+renderAsset
+    :: SNetworkId n
+    -> (Address, (TokenBundle, [(String, String)]))
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
    -> (Address, (TokenBundle, AssetMetadata))
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    -> Value
+renderAsset n (addr, (bundle, metadata)) =
+    object
+        [ "address" .= renderAddress n addr
+        , "bundle" .= renderBundle bundle
+        , "metadata" .= renderMetadata metadata
+        ]
+
+parseMetadata :: Value -> Parser [(String, String)]
+parseMetadata = parseJSON >=> mapM parseMetadataValue
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
parseAssetMetadata :: Value -> Parser AssetMetadata
parseAssetMetadata = parseJSON >=> mapM parseMetadataValue
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+renderAsset
+    :: SNetworkId n
+    -> (Address, (TokenBundle, [(String, String)]))
+    -> Value
+renderAsset n (addr, (bundle, metadata)) =
+    object
+        [ "address" .= renderAddress n addr
+        , "bundle" .= renderBundle bundle
+        , "metadata" .= renderMetadata metadata
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
        , "metadata" .= renderAssetMetadata metadata
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+parseAssets n = parseJSON >=> mapM (parseAsset n)
+
+parseAsset
+    :: SNetworkId n
+    -> Value
+    -> Parser (Address, (TokenBundle, [(String, String)]))
+parseAsset n = withObject "Asset" $ \o -> do
+    addr <- o .: "address" >>= parseAddress n
+    bundle <- o .: "bundle" >>= parseBundle
+    metadata <- o .: "metadata" >>= parseMetadata
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
    metadata <- o .: "metadata" >>= parseAssetMetadata
```

File: lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+parseMetadata = parseJSON >=> mapM parseMetadataValue
+
+parseMetadataValue :: Value -> Parser (String, String)
+parseMetadataValue = withObject "MetadataValue" $ \o -> do
+    k <- o .: "key"
+    v <- o .: "value"
+    pure (k, v)
+
+renderMetadata :: [(String, String)] -> Value
+renderMetadata =
```
Comment (HeinrichApfelmus) [inline lib/local-cluster/lib/Cardano/Wallet/Launch/Cluster/Http/Faucet/SendFaucetAssets.hs]:
```suggestion
renderAssetMetadata :: AssetMetadata -> Value
renderAssetMetadata =
```

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent, thank you! 😊 

A couple of suggestion to make the use of `AssetMetadata` more consistent, but otherwise, good to go.

**Pull request merged**